### PR TITLE
TLS Refactoring

### PR DESCRIPTION
**Pull request by orozery**
This PR makes the following changes:
- Move peer server TLS termination from the controlplane to the dataplane.
- Change tunnel establishment method from POST to CONNECT.
- Peer authorization token returned using a header instead of a json-encoded http body.
- Change controlplane ext_authz server from HTTP go gRPC.
- Remove SNI proxies in both the controlplane and the dataplane.
- Support dynamic peer certificates in the dataplane via SDS from controlplane.
- Support dynamic peer certificates in the controlplane (not yet tested).
- Remove peer CA certificate, and instead use a new "site CA" for signing controlplane and dataplane certificates (used for controlplane<->dataplane traffic).

### CODE REVIEW

File: pkg/dataplane/server/server.go [hunk 1]
```diff
@@ -41,66 +37,113 @@ func (d *Dataplane) StartDataplaneServer(dataplaneServerAddress string) error {
 		WriteTimeout:      2 * time.Second,
 		IdleTimeout:       120 * time.Second,
 		MaxHeaderBytes:    10 * 1024,
-		TLSConfig:         d.parsedCertData.ServerConfig(),
+		TLSConfig: &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			ClientAuth: tls.RequireAndVerifyClientCert,
+			GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+				// this function is defined for the sake of skipping certificate file reading
+				// in net/http.Server.ServeTLS
```
Comment (elevran) [inline pkg/dataplane/server/server.go:45]: document where the certificate is coming from...?

**Review by kfirtoledo — COMMENTED** [review]
LGTM, small comments

**Review by elevran — APPROVED** [review]

File: cmd/cl-controlplane/app/server.go [hunk 2]
```diff
@@ -49,20 +48,21 @@ const (
...
 	KeyFile = "/etc/ssl/private/clink-controlplane.pem"
 
-	// httpServerAddress is the address of the localhost HTTP server.
-	httpServerAddress = "127.0.0.1:1100"
-	// grpcServerAddress is the address of the localhost gRPC server.
-	grpcServerAddress = "127.0.0.1:1101"
+	// PeerTLSDirectory is the path to the directory holding the peer TLS certificates.
+	PeerTLSDirectory = "/etc/ssl/certs/clink"
+	// PeerCertificateFile is the name to the peer certificate file.
+	PeerCertificateFile = "cert.pem"
```
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/cl-controlplane/app/server.go:61]: The controlplane certificate is not used for inter-cluster communication. It should not be the same as the peer certificate.

[the remark this answers is not in the document]
Comment (orozery) [reply cmd/cl-controlplane/app/server.go:61]: The controlplane<->(remote controlplane) communcation is using the peer certificate.
The controlplane certificate is only used for controlplane<->dataplane.

File: cmd/cl-go-dataplane/app/server.go [hunk 3]
```diff
@@ -73,26 +74,36 @@ func (o *Options) RequiredFlags() []string {
...
-func (o *Options) runGoDataplane(peerName, dataplaneID string, parsedCertData *tls.ParsedCertData) error {
+func (o *Options) runGoDataplane(dataplaneID string, parsedCertData *tls.ParsedCertData) error {
 	controlplaneTarget := net.JoinHostPort(o.ControlplaneHost, strconv.Itoa(cpapi.ListenPort))
 
-	logrus.Infof("Starting go dataplane, Name: %s, ID: %s", peerName, dataplaneID)
+	logrus.Infof("Starting go dataplane, ID: %s", dataplaneID)
+
+	controlplaneClient, err := grpc.NewClient(
+		controlplaneTarget,
+		grpc.WithTransportCredentials(credentials.NewTLS(parsedCertData.ClientConfig("cl-controlplane"))),
```
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/cl-go-dataplane/app/server.go:84]: "cl-controlplane" is the expected server name of the controlplane (as seen by the dataplane).
This TLS config is used for mTLS (CA for server verification + client cert for gaining trust from the controlplane)

File: cmd/clusterlink/config/config.go [hunk 4]
```diff
@@ -33,6 +33,8 @@ const (
 	ControlplaneDirectoryName = "controlplane"
 	// DataplaneDirectoryName is the directory name containing dataplane server configuration.
 	DataplaneDirectoryName = "dataplane"
+	// CADirectoryName is the directory name containing site CA configuration.
+	CADirectoryName = "ca"
```
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/clusterlink/config/config.go:37]: The directory structure is as follows:
<fabric_name>/
<fabric_name>/<peer name>
<fabric_name>/<peer name>/<peer cert and private key files>
<fabric_name>/<peer name>/ca/<site ca cert and private key file, used for controlplane<->dataplane verficiation>
<fabric_name>/<peer name>/controlplane/<controlplane cert and private key file>
<fabric_name>/<peer name>/dataplane/<dataplane cert and private key file>

File: pkg/controlplane/api/authz.go [hunk 5]
```diff
@@ -39,6 +35,9 @@ const (
 	// TargetClusterHeader holds the name of the target cluster.
 	TargetClusterHeader = "host"
 
+	// AccessTokenHeader holds the access token for an exported service, sent back by the server.
+	AccessTokenHeader = "x-access-token"
```
[the remark this answers is not in the document]
Comment (orozery) [reply pkg/controlplane/api/authz.go:39]: Notice this is a header the server returns to the client (and not a client sending to server). Still makes sense?

File: pkg/controlplane/authz/server.go [hunk 6]
```diff
@@ -35,141 +41,213 @@ type server struct {
...
-	if ip == "" {
-		http.Error(w, fmt.Sprintf("missing '%s' header", api.ClientIPHeader), http.StatusBadRequest)
-		return
+// check an egress dataplane connection.
+func (s *server) checkEgress(ctx context.Context, req *authv3.CheckRequest) *authv3.CheckResponse {
+	httpReq := req.Attributes.Request.Http
+	headers := httpReq.Headers
+
+	expectedHeaders := []string{api.ClientIPHeader, api.ImportNameHeader, api.ImportNamespaceHeader}
+	for _, header := range expectedHeaders {
```
[the remark this answers is not in the document]
Comment (orozery) [reply pkg/controlplane/authz/server.go:77]: The ext_authz filter passes information on an HTTP request.
Maybe you can use an additional filter to conver the HTTP headers to metadata:
https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/header_to_metadata_filter
But I don't see any advantage in that. It just complicates things.

File: pkg/dataplane/server/dataplane.go [hunk 7]
```diff
@@ -107,24 +127,119 @@ func (d *Dataplane) GetListeners() map[string]*listener.Listener {
...
+	if trustedCa == nil {
+		return fmt.Errorf("no trusted CA")
+	}
+
+	caBytes := trustedCa.GetInlineBytes()
+	if caBytes == nil {
+		return fmt.Errorf("no CA bytes embedded")
+	}
+
+	caCertPool := x509.NewCertPool()
```
[the remark this answers is not in the document]
Comment (orozery) [reply pkg/dataplane/server/dataplane.go:198]: The code assumes a single SDS secret which holds all CAs.
We can encode multiple CAs to a single secret on the controlplane side (the one that creates the SDS secrets).

File: cmd/cl-controlplane/app/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (elevran) [reply cmd/cl-controlplane/app/server.go:61]: I understand this is the design, so clarified the working of SPIRE to ensure it is aligned.
SPIRE will create only one certificate for the control plane. It should be used for both control-remote control and for control-data planes. Or we would need to still manage our own certificates for one of these channels.
An option is to rely on SPIRE/clusterlink CLI for the peer to peer only and locally create (self signed) certs and CA for control and data plane communications in the control plane (data plane would get those over SDS if needed).

File: cmd/clusterlink/config/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (elevran) [reply cmd/clusterlink/config/config.go:37]: See earlier comment on self signed for in cluster vs peer to peer.
perhaps we can rethink this so control-data are generated on the fly (in CP or operator)?

File: cmd/cl-controlplane/app/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/cl-controlplane/app/server.go:61]: > An option is to rely on SPIRE/clusterlink CLI for the peer to peer only and locally create (self signed) certs and CA for control and data plane communications in the control plane (data plane would get those over SDS if needed).

This is already implemented in this PR.

File: pkg/controlplane/api/authz.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (elevran) [reply pkg/controlplane/api/authz.go:39]: agree. See this [stack overflow](https://stackoverflow.com/questions/51503024/how-does-server-return-jwt-token-to-the-client) for discussion/reference. Header is fine as it is a "private protocol"

File: cmd/clusterlink/config/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/clusterlink/config/config.go:37]: Let's talk about it the next time we meet.

File: cmd/cl-controlplane/app/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (elevran) [reply cmd/cl-controlplane/app/server.go:61]: ok. I did not realize that.
SPIRE would still allocate a certificate for the data plane Pod. Will we use it?

File: cmd/clusterlink/config/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (elevran) [reply cmd/clusterlink/config/config.go:37]: I guess I'm not clear on who/when we create these (CLI, operator, control plane when started...)

File: cmd/cl-controlplane/app/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/cl-controlplane/app/server.go:61]: The certificate allocated by SPIRE to the controlplane pod will be used for inter-cluster communication (peer-to-peer).
The certificates for intra-cluster communcation (including the dataplane pod) will be generated by us when deploying the site.

File: cmd/clusterlink/config/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (orozery) [reply cmd/clusterlink/config/config.go:37]: All of it is created in the CLI `clusterlink create peer`

[the remark this answers is not in the document]
Comment (orozery) [reply cmd/clusterlink/config/config.go:37]: Except of course for `<fabric_name>` which is created by `clusterlink create fabric`

**Pull request merged**